Source/TrustEvaluator.swift

```diff
@@ -0,0 +1,102 @@
+// copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+// SOFTWARE.
+//
+
+import Foundation
+
+internal protocol TrustEvaluatorType: class {
+    var pinnables: [Pinnable] { get set }
+    var strategy: ValidationStrategy { get set }
+    
+    func handleChallenge(_ challenge: URLAuthenticationChallengeType, completion: @escaping (URLSession.AuthChallengeDisposition, URLCredential?) -> Void)
+}
+
+final internal class TrustEvaluator: TrustEvaluatorType {
+    
+    var strategy: ValidationStrategy
+    lazy var pinnables: [Pinnable] = {
+        return self.strategy.pinnables(from: PEMCertificate.pinnedCertificates)
+    }()
+    
+    required init(system: SystemType) {
+        switch system.device.systemVersion.compare("10.3", options: .numeric) {
+        case .orderedAscending:
+            strategy = .certificates
+        default:
+            strategy = .publicKeys
+        }
+    }
+    
+    func handleChallenge(_ challenge: URLAuthenticationChallengeType, completion: @escaping (URLSession.AuthChallengeDisposition, URLCredential?) -> Void) {
+        let space = challenge.getProtectionSpace()
+        guard shouldPinChallenge(challenge),
+            let trust = space.serverTrust else {
+            completion(.performDefaultHandling, nil)
+            return
+        }
+        
+        guard validate(trust, host: space.host) else {
+            completion(.cancelAuthenticationChallenge, nil)
+            return
+        }
+        
+        completion(evaluate(trust), nil)
+    }
+    
+    private func shouldPinChallenge(_ challenge: URLAuthenticationChallengeType) -> Bool {
+        let space = challenge.getProtectionSpace()
+        guard space.authenticationMethod == NSURLAuthenticationMethodServerTrust,
+            space.host.contains(".usebutton.com") else {
+                return false
+        }
+        return true
+    }
+    
+    private func validate(_ trust: SecTrust, host: String) -> Bool {
+        let policy = SecPolicyCreateSSL(true, host as CFString)
+        SecTrustSetPolicies(trust, policy)
+        
+        var result = SecTrustResultType.invalid
+        let status = SecTrustEvaluate(trust, &result)
+        guard status == errSecSuccess,
+            result == .proceed || result == .unspecified else {
+                return false
+        }
+        return true
+    }
+    
+    private func evaluate(_ trust: SecTrust) -> URLSession.AuthChallengeDisposition {
+        let count = SecTrustGetCertificateCount(trust)
```

Comment:
The documentation for this method recommends calling `SecTrustGetTrustResult` prior to this call. Is that necessary here?

For reference:
```
/*!
    @function SecTrustGetCertificateCount
    @abstract Returns the number of certificates in an evaluated certificate
    chain.
    @param trust A reference to a trust object.
    @result The number of certificates in the trust chain, including the anchor.
    @discussion Important: if the trust reference has not yet been evaluated,
    this function will evaluate it first before returning. If speed is critical,
    you may want to call SecTrustGetTrustResult first to make sure that a
    result other than kSecTrustResultInvalid is present for the trust object.
 */
```

Reply:
That's done in the `validate` function.